Add adaptive pooling (1D, 2D, 3D) support across JAX, NumPy, TensorFlow, and PyTorch backends

MalyalaKarthik66 opened the pull request:
fix #21813 

# Add adaptive pooling support across major backends
This PR implements adaptive average and max pooling for 1D, 2D, and 3D across the JAX, NumPy, TensorFlow, and PyTorch backends.

- For PyTorch, native adaptive pooling ops are used.

- For JAX and TensorFlow, adaptive pooling is implemented using an efficient n-dimensional two-pool gather algorithm, eliminating multiple for-loops and providing robust performance on CPU, GPU, and TPU.

- For NumPy, adaptive pooling is implemented using the same n-dimensional two-pool gather algorithm with NumPy stride tricks, providing an efficient pure-NumPy implementation for CPU.

- All corresponding unit tests for JAX, NumPy, TensorFlow, and PyTorch adaptive pooling pass successfully.

- Verified in real training model tests — both TensorFlow and PyTorch pass on GPU and CPU environments.

hertschuh commented on keras/src/backend/jax/__init__.py:
```diff
@@ -25,6 +25,8 @@
 from keras.src.backend.jax.core import stop_gradient
 from keras.src.backend.jax.core import vectorized_map
+from keras.src.backend.jax.nn import adaptive_avg_pool
+from keras.src.backend.jax.nn import adaptive_max_pool
```
Revert this file.

hertschuh commented on keras/src/backend/jax/nn.py:
```diff
@@ -1464,3 +1464,368 @@ def _pair(x):
+    return small_window, big_window
+
+
+def compute_static_gather_indices(input_dim, output_size, big_window):
```
Rename `_compute_static_gather_indices` to distinguish this from other ops.

MalyalaKarthik66 commented:
@hertschuh 
I’ve updated the PR with all the requested changes. I introduced shared base classes for adaptive pooling and refactored the 1D/2D/3D layers accordingly, and also implemented the NumPy backend version.

I initially moved `compute_adaptive_pooling_window_sizes` into `ops/operation_utils`, but that created a circular import since `ops/__init__` imports backend symbols. I’ve moved the utility into `backend/common/backend_utils.py` instead and import it from each backend.

Please let me know if anything else needs to be adjusted — happy to update further.

hertschuh commented on keras/src/backend/openvino/nn.py:
```diff
@@ -145,6 +153,14 @@ def average_pool(
+    """Adaptive average pooling - OpenVINO backend not yet supported."""
+    raise NotImplementedError(
+        "Adaptive pooling not implemented for OpenVINO. "
+        "Use JAX or Torch backend."
```
Same comment.

hertschuh commented on keras/src/backend/jax/nn.py:
```diff
@@ -1464,3 +1467,398 @@ def _pair(x):
+
+
+def adaptive_avg_pool(inputs, output_size, data_format="channels_first"):
+    dims = inputs.ndim - 2
```
The code is good, but can you move it to follow the correct order, the same way you did for the Tensorflow backend.

hertschuh commented on keras/src/layers/pooling/adaptive_max_pooling1d.py:
```diff
@@ -0,0 +1,84 @@
+            raise TypeError(
+                "`output_size` must be an integer. Received output_size={} "
+                "of type {}".format(output_size, type(output_size))
+            )
```
+1

hertschuh commented on keras/src/ops/nn.py:
```diff
@@ -1162,6 +1163,58 @@ def max_pool(
+    """
+    if data_format is None:
+        data_format = config.image_data_format()
+    return backend.nn.adaptive_max_pool(
```
You'll need a class called `AdaptiveMaxPool`. Look at `AveragePool` below for an example.

Then here, add:
```python
 if any_symbolic_tensors((inputs,)):
        return AdaptiveMaxPool(output_size, data_format).symbolic_call(inputs)
```

hertschuh commented on keras/src/layers/pooling/adaptive_average_pooling1d.py:
```diff
@@ -0,0 +1,58 @@
+            raise TypeError(
+                f"`output_size` must be an integer. "
+                f"Received: {output_size} of type {type(output_size)}"
+            )
```
For consistency, also allow a tuple of size 1 and mention it in the error message.

hertschuh commented on keras/src/layers/pooling/adaptive_max_pooling1d.py:
```diff
@@ -0,0 +1,58 @@
+            raise TypeError(
+                "`output_size` must be an integer. Received output_size={} "
+                "of type {}".format(output_size, type(output_size))
+            )
```
For consistency, also allow a tuple of size 1 and mention it in the error message.

hertschuh commented:
Great! This is ready to go except:
- can you rebase (one file need merging)?
- can you run the code formatter?
